### DISCUSSION TITLE
Showing nicer tables in plugin

### DISCUSSION THREAD

**Discussion by markkuleinio** [netbox-community/netbox#15525] [Help Wanted!] [answered]
In my quest to display arbitrary external data in a NetBox view, I've got this far:

![image](https://github.com/netbox-community/netbox/assets/25136274/ff84f526-4dfb-4ab8-961b-bcb592bad3da)

Meaning that I can display data that is from a static dict for now. There is even pagination (`views.py`):

``` python
        table = MyDataTable(my_precious_data_as_dict)
        paginate = {
            'paginator_class': EnhancedPaginator,
            'per_page': get_paginate_count(request),
        }
        tables.RequestConfig(request, paginate).configure(table)
        return render(request, "my_plugin/my_data_list.html", { "table": table})
```

Naturally I'd like to get the list a bit more like NetBox-looking. How is that actually achieved?

My table is simply (`tables.py`):

```python
import django_tables2 as tables

class MyDataTable(tables.Table):
    name = tables.Column()
    value = tables.Column()
```

I figured out that I cannot use `NetBoxTable` or `BaseTable` since they are model-based (and complain about missing `_meta`).

The template is (`my_data_list.html`):

```html
{% extends 'base/layout.html' %}
{% load render_table from django_tables2 %}

{% block title %}My Data{% endblock %}

{% block content %}
<div class="row">
    <div class="col-md-9">
        {% render_table table %}
    </div>
</div>
{% endblock content %}
```

`base/layout.html` is the only I've yet found that works with dict data, others (like `generic/object_list.html`) require "proper" objects and fail in trying to access the missing attributes.

Any hints how to improve the layout and get the NetBox look and feel? Should I just to skip `render_table`, use a `for` loop to show the data, and reverse-engineer HTML output from other pages? This Django templating ideology is new for me.

**Comment by wouterdebruijn** [comment, marked as answer]
Because the data provided isn't linked to any NetBox model, we aren't able the extend the NetboxTable class, which would normally set the required classes. However, we can set those classes manually.

```python
class MyDataTable(tables.Table):
    name = tables.Column()
    value = tables.Column()

    class Meta:
        attrs = {
            'class': 'table table-hover',
        }
```

This table is still quite basic compared to NetBox tables, where sorting, filtering, etc is allowed. Those functions are all implemented in NetBoxTable.

**Reply by jeremystretch** [reply]
Most of the dynamic table functionality comes from the BaseTable class from which NetBoxTable inherits. It's not yet officially part of the plugins API but can be used without much risk.

**Reply by markkuleinio** [reply]
Adding the attrs made wonders already! Thank you!

`tables.Table` already provides sorting, it's nice. I don't mind filtering missing.

Any further hints to get the pagination links to be something better that the current plain "12next"?

**Reply by markkuleinio** [reply]
Currently:

<img width="1119" alt="image" src="https://github.com/netbox-community/netbox/assets/25136274/49504034-ec28-4fff-a883-6de73636ec2e">

<img width="625" alt="image" src="https://github.com/netbox-community/netbox/assets/25136274/9f296592-d879-4ad9-b5d0-aaa36d7c8a73">

**Reply by markkuleinio** [reply]
The pagination link also only shows below the data even though the user preference for it is set to Both.

**Reply by markkuleinio** [reply]
With this I got forward: there are better pagination buttons, but they don't work yet:

```html
{% extends 'base/layout.html' %}

{% block title %}My Data{% endblock %}

{% block content %}
{% include 'htmx/table.html' %}
{% endblock content %}
```
(found this from `generic/object_list.html`)

![image](https://github.com/netbox-community/netbox/assets/25136274/fce5dfad-c755-40b0-ae9c-deec91c9a2c6)

Also, sorting stopped working with this

**Reply by markkuleinio** [reply]
Wrapping the table like this made sorting and paginating work:

```
<div class="htmx-container">
{% include 'htmx/table.html' %}
</div>
```

**Comment by markkuleinio** [comment]
So the final solution for this exercise:

`tables.py`:
``` Python
import django_tables2 as tables


class MyDataTable(tables.Table):
    name = tables.Column()
    value = tables.Column()

    class Meta:
        attrs = {
            "class": "table table-hover",
        }
```

`views.py`:
``` Python
import django_tables2 as tables
from utilities.htmx import is_embedded, is_htmx
from utilities.paginator import EnhancedPaginator, get_paginate_count

...

        table = MyDataTable(my_precious_data_as_dict)
        paginate = {
            'paginator_class': EnhancedPaginator,
            'per_page': get_paginate_count(request),
        }
        tables.RequestConfig(request, paginate).configure(table)
        if is_htmx(request):
            if is_embedded(request):
                table.embedded = True
            return render(request, 'htmx/table.html', {'table': table})
        return render(request, "my_plugin/my_data_list.html", { "table": table})
```

`my_data_list.html`:
``` html
{% extends 'base/layout.html' %}

{% block title %}My Data{% endblock %}

{% block content %}
<div class="htmx-container">
{% include 'htmx/table.html' %}
</div>
{% endblock content %}

```

To be clear and honest: I have no idea how these will survive NetBox major/minor version upgrades 😁

**Reply by markkuleinio** [reply]
As expected, changes will be needed for NetBox 4.0: `is_htmx()` and `is_embedded()` are not there anymore, so at least these changes will be needed:

```python
from utilities.htmx import htmx_partial
...
        if htmx_partial(request):
            if not request.htmx.target:
                table.embedded = True
```
